Migrating to the Auth0 Email Search API

kopertop opened the pull request:
This post probably shouldn't be published until my fork (or another patch) is updated for the Node.js SDK. I included a note about that which should be deleted after an official update is released.

It was suggested by Ado Kukic that I submit this blog post to Auth0 first. I will then cross-post on my blog with a link to Auth0 after any changes are made.

brunokrebs commented on _posts/2017-10-25-migrating-to-email-api.markdown:
```diff
@@ -0,0 +1,291 @@
+---
+layout: post
+title: "Migrating to the new Email Search API"
```
I would add Auth0 before "Email Search API"

brunokrebs commented on _posts/2017-10-25-migrating-to-email-api.markdown:
```diff
@@ -0,0 +1,291 @@
+---
+layout: post
+title: "Migrating to the new Email Search API"
+description: "If you use the Search API to map users by an email address, it's time to move to the new Email Search API"
```
Same here...

brunokrebs replied:
"users by an email address" sounds contradictory. I think it would be better to have "users by email addressES" or "a user by an email address".

kukicado replied:
"If you are using the Search API to map users by email addresses, it's time to move to the new Auth0 Email Search API"

brunokrebs commented on _posts/2017-10-25-migrating-to-email-api.markdown:
```diff
@@ -0,0 +1,291 @@
+- New Features
+---
+
+One very common use case for the Auth0 Search API is to identify users by email address, merging accounts together, or linking existing accounts based on a shared email address. In the case of my company, Cannabiz Media, we also used this API to identify new users and create accounts when a user signed up, de-duplicating by email address to prevent users from having multiple accounts.
```
You start the paragraph with "One very common use case for the Auth0 Search API is to" and then you list three possibilities:

- identify users by email address
- merging accounts together
- linking existing accounts based on a shared email address

As these possibilities are related to the same sentence, they should keep the same form. That is, instead of "merging accounts", it should be "merge accounts". The same goes for the last option, instead of "linking existing accounts" it should be "link existing accounts".

brunokrebs replied:
You can add a link to your company's website.

kukicado replied:
I think all three items make up the use case Bruno. So what I would change the start to is "A common use case.."

brunokrebs commented on _posts/2017-10-25-migrating-to-email-api.markdown:
```diff
@@ -0,0 +1,291 @@
+
+One very common use case for the Auth0 Search API is to identify users by email address, merging accounts together, or linking existing accounts based on a shared email address. In the case of my company, Cannabiz Media, we also used this API to identify new users and create accounts when a user signed up, de-duplicating by email address to prevent users from having multiple accounts.
+
+## Introduction to the Search API
```
Search API Introduction

kukicado replied:
I would leave this as is :)

brunokrebs commented on _posts/2017-10-25-migrating-to-email-api.markdown:
````diff
@@ -0,0 +1,291 @@
+
+## Introduction to the Search API
+
+The search API is a generic API for searching Users, including with the ability to search for any generic metadata associated with a user's ```app_metadata``` or ```user_metadata```. This means if you had something like:
````
Users -> users (no need for capital case)

brunokrebs replied:
Actually I would replace the first sentence with:

    The search API is a generic API for searching users that includes the ability to search for any generic metadata associated with a user's (```app_metadata``` or ```user_metadata```).

brunokrebs replied:
"This means if you had something like" -> "For example, if you had something like"

brunokrebs commented on _posts/2017-10-25-migrating-to-email-api.markdown:
````diff
@@ -0,0 +1,291 @@
+you could easily search for any users on the Trial plan by doing a search like:
+
+```
+app_metdata.plan:Trial
````
missing an a in metadata

brunokrebs commented on _posts/2017-10-25-migrating-to-email-api.markdown:
````diff
@@ -0,0 +1,291 @@
+app_metdata.plan:Trial
+```
+
+That's great for general searches, but it means that Auth0 has to store a lot of indexes, including a lot for things people never search for. In fact, most of the uses for the search API were just for looking up users by email. If you've ever set up a Search system, you know how ineficient it is to have unused indexes, and what kind of impact that has to search.
````
"including a lot for things people never search for" -> "including a lot **of** things people never search for"

kukicado replied:
"set up a Search system" => "set up a search system"

brunokrebs commented on _posts/2017-10-25-migrating-to-email-api.markdown:
```diff
@@ -0,0 +1,291 @@
+
+### Why using the Search API for authentication workflow is bad
+
+The search API isn't designed for high availability. Over the past week, I've noticed at least 3 occasions when it had issues (either complete outages, or just a partial outage). Your system is only as strong as it's weakest link, so if your login requires access to a system that's frequently unavailable, your entire system is effectively broken.
```
"isn't designed for high availability" -> It isn't? Even though this is true, it might sound bad for Auth0...

brunokrebs replied:
"Over the past week, I've noticed at least 3 occasions when it had issues (either complete outages, or just a partial outage)" -> same here, perhaps we should change wording...

kukicado replied:
I agree with Bruno here.. can we change to something like:

"The current Search API, while giving a lot of flexibility to developers, isn't ideal for all use cases. Email lookup is one such case. User data accessed via the Search API typically takes longer to return due to the nature of the system that gives us the flexibility to store and look up any type of data, so creating a dedicated endpoint for accessing user email information is a step in the right direction." or something similar?

kopertop replied:
I want to make it very clear that the Search API shouldn't be used for High Availability requirements... how about:

> The current Search API, while giving a lot of flexibility to developers, isn't ideal for high-availability systems.

brunokrebs commented on _posts/2017-10-25-migrating-to-email-api.markdown:
```diff
@@ -0,0 +1,291 @@
+
+### Use-cases for mapping by email
+
+Auth0 allows you to have multiple identities (authentication mechanisms) for the same account. For example, you could set up a Database (username/password) account that also is connected to your Gmail login, no matter which login you use, you'll be connected to the same account.
```
"to have multiple identities" -> "to have multiple identity providers"

brunokrebs replied:
"For example, you could set up a Database (username/password) account that also is connected to your Gmail login, no matter which login you use, you'll be connected to the same account." -> "For example, you could set up an external database connection alongside with Google social login and merge profiles by their email addresses. Like that, no which login mechanism users choose, they'll be connected to the same account."

kukicado replied:
👍

brunokrebs commented on _posts/2017-10-25-migrating-to-email-api.markdown:
```diff
@@ -0,0 +1,291 @@
+
+Auth0 allows you to have multiple identities (authentication mechanisms) for the same account. For example, you could set up a Database (username/password) account that also is connected to your Gmail login, no matter which login you use, you'll be connected to the same account.
+
+In fact, the recommended rule to [Link accounts by email address](https://github.com/auth0/rules/blob/master/rules/link-users-by-email.md) used the Search API to do this.
```
"used the Search..." -> "uses the Search..."

kopertop replied:
Actually the rule was modified, and no longer does use that.

brunokrebs commented on _posts/2017-10-25-migrating-to-email-api.markdown:
```diff
@@ -0,0 +1,291 @@
+
+It's a common request, a user logs in with their Google account and they get migrated right into their normal account.
+
+### The Search API isn't designed for maximum availability
```
change wording?

brunokrebs commented on _posts/2017-10-25-migrating-to-email-api.markdown:
```diff
@@ -0,0 +1,291 @@
+
+The Email Lookup API doesn't use the search system, with it's thousands of different indexes. Instead, it uses a very simple index on the primary data store, the same DB that has had 99.99% uptime over the past 6 months. Using this API in your authentication workflow is completely safe.
+
+### What code to replace, and how to migrate from the Search API
```
big section title, huh? Perhaps "Migrating to Email Search API"?

brunokrebs commented on _posts/2017-10-25-migrating-to-email-api.markdown:
````diff
@@ -0,0 +1,291 @@
+
+### What code to replace, and how to migrate from the Search API
+
+In order to make the migration, there's only a few lines in the original rule we need to replace. Instead of hitting ```/users```, we need to hit ```/users-by-email```:
````
"in the original rule we need to replace" -> "in the original rule **that** we need to replace"

brunokrebs replied:
no need to triple \` just one in the beginning and one in the end is enough. That is, for inline code like this.

brunokrebs commented on _posts/2017-10-25-migrating-to-email-api.markdown:
````diff
@@ -0,0 +1,291 @@
+  },
+```
+
+Unlike with the search API, this API doesn't allow additional filters such as excluding the current user, so we also need to add a filter to our data:
````
search API -> Search API (Capital S to keep consistent)

brunokrebs replied:
", so we also need to add a filter to our data" -> "Therefore, we also need to..."

brunokrebs commented on _posts/2017-10-25-migrating-to-email-api.markdown:
```diff
@@ -0,0 +1,291 @@
+
+Other than that, the data returned is identical so there's no problem with the remainder of our rule.
+
+### Limitations, things that it can not be used for
```
"### Email Search API Limitations"

this gives the blog post more SEO points when searching for "Email Search API"

brunokrebs commented on _posts/2017-10-25-migrating-to-email-api.markdown:
````diff
@@ -0,0 +1,291 @@
+
+### Limitations, things that it can not be used for
+
+As mentioned briefly above, there's no additional filters allowed via the ```users-by-email``` API, so any additional filtering needs to be done client-side. You also can't search for anything in the ```app_metadata```, so if you rely on rules to match things like by a ```customer_id``` or other ID outside of the email address, you'll still need to rely on the Search API. However, keep in mind th ...
````
"you'll still need to rely on the Search API" -> "you'll still need the Search API"

brunokrebs replied:
"However, keep in mind that anything that relies on the Search API will not be available 99.99% of the time." -> "However, keep in mind that anything that relies on the Search API won't be as available as the Email Search API."

brunokrebs commented on _posts/2017-10-25-migrating-to-email-api.markdown:
````diff
@@ -0,0 +1,291 @@
+
+As mentioned briefly above, there's no additional filters allowed via the ```users-by-email``` API, so any additional filtering needs to be done client-side. You also can't search for anything in the ```app_metadata```, so if you rely on rules to match things like by a ```customer_id``` or other ID outside of the email address, you'll still need to rely on the Search API. However, keep in mind th ...
+
+## How Cannabiz Media uses the new Email Lookup API
````
Cannabiz Media and the Email Search API

kukicado replied:
I think either one works :)

brunokrebs replied:
yep, I just try to make shorter titles as our content area is not that wide

brunokrebs commented on _posts/2017-10-25-migrating-to-email-api.markdown:
```diff
@@ -0,0 +1,291 @@
+
+## How Cannabiz Media uses the new Email Lookup API
+
+Of course, linking users by email address is the most common use-case and essentially why this new API endpoint was created. However, there are other reasons to need to look up a user by email address. At my company, [Cannabiz Media](https://cannabiz.media), we use WooCommerce as our Storefront, which automatically creates user accounts in Auth0 when a user purchases a new subscription. A WooComm ...
```
"and essentially why this new API endpoint was created." -> "and essentially this is why new API endpoint was created."

brunokrebs replied:
"However, there are other reasons to need to look up a user by email address." -> "However, there are other reasons where looking up users by email addresses is needed."

brunokrebs commented on _posts/2017-10-25-migrating-to-email-api.markdown:
````diff
@@ -0,0 +1,291 @@
+}
+```
+
+As I mentioned, this webhook would be fired if the subscription was added, updated, or cancelled, so we couldn't assume that just because the webhook fired this was a brand-new user. Obviously there's no Auth0 ID there, so we used the ```billing.email``` to see if there already was a user for this account, or if we needed to create a new one.
````
"As I mentioned, this webhook would be fired if the subscription was added, updated, or cancelled, so we couldn't assume that just because the webhook fired this was a brand-new user."
->
"As the webhook is fired when subscriptions are added, updated, and cancelled, we can't assume that these events always reference brand-new users."

brunokrebs commented on _posts/2017-10-25-migrating-to-email-api.markdown:
```diff
@@ -0,0 +1,291 @@
+
+### Making the Switch to the Users-By-Email API
+
+The search API had issues, and one of the problems we experienced was not being able to add new subscribers, or modify subscribers, when the Search API was down. We had users subscribing while the Search API was down, and we were getting angry calls from our customers wondering why they couldn't log in yet. Worst of all, before we double-checked in DynamoDB, we ended up with *duplicate accounts*, ...
```
search API -> Search API

brunokrebs commented on _posts/2017-10-25-migrating-to-email-api.markdown:
```diff
@@ -0,0 +1,291 @@
+
+## Conclusion
+
+Auth0's Search API uses ElasticSearch. Anyone who's ever dealt with ElasticSearch knows how complicated it can be to get it right. What makes Auth0's situation even more complex is that they're using it to host *customers* data, which can be any random format, not something that adhears to a strict format. Auth0 knows this, and they've come up with a good solution for 90% of the use-cases for sea ...
```
adhears -> adheres

brunokrebs commented on _posts/2017-10-25-migrating-to-email-api.markdown:
```diff
@@ -0,0 +1,291 @@
+
+### Email Search is more limitted, but fits the general use case
+
+Email Searching doesn't solve all of the problems. Auth0 will still need to work on making the Search API more resiliant and reliable for developers, but it's no longer a critical issue. The issue identified fits our use case, and probably will fit yours too. It might take a little bit more work to get into your workflow, but it's well worth it to ensure users have a good experience.
```
"Email Searching doesn't" -> "Email Search API doesn't"

brunokrebs commented on _posts/2017-10-25-migrating-to-email-api.markdown:
```diff
@@ -0,0 +1,291 @@
+
+Email Searching doesn't solve all of the problems. Auth0 will still need to work on making the Search API more resiliant and reliable for developers, but it's no longer a critical issue. The issue identified fits our use case, and probably will fit yours too. It might take a little bit more work to get into your workflow, but it's well worth it to ensure users have a good experience.
+
+For new developers just starting out, the new API is actually simpler. Instead of having to craft a special query, you just hit an endpoint.
```
"For new developers just starting out" -> "For developers just starting out" (redundant to say new developers just starting)

brunokrebs commented on _posts/2017-10-25-migrating-to-email-api.markdown:
```diff
@@ -0,0 +1,291 @@
+
+### There are more use-cases than just the Account-linking rule
+
+As you can see from our use-case, looking up a user by email is useful to a lot more applications then just an Auth0 rule to merge accounts. In fact, we're currently using it to help us with account creation, and we're using it to link users from Auth0 to our Analytics platform. The new API should also be significantly less resource-intensive for Auth0, and with fewer people using the Search API  ...
```
"to our Analytics platform" -> "to our analytics platform" (no need to capital A as Analytics is not a proper noun)

brunokrebs replied:
"these frequent request" -> "these frequent requests" (multiple requests not singular)

brunokrebs commented on _posts/2017-10-25-migrating-to-email-api.markdown:
```diff
@@ -0,0 +1,291 @@
+
+As you can see from our use-case, looking up a user by email is useful to a lot more applications then just an Auth0 rule to merge accounts. In fact, we're currently using it to help us with account creation, and we're using it to link users from Auth0 to our Analytics platform. The new API should also be significantly less resource-intensive for Auth0, and with fewer people using the Search API  ...
+
+Yes, search still needs to be fixed, but this API should reduce the impact search has on your workflow.
```
Fixed? Perhaps improved?

kukicado replied:
I would remove this last sentence altogether.

kukicado commented on _posts/2017-10-25-migrating-to-email-api.markdown:
````diff
@@ -0,0 +1,291 @@
+}
+```
+
+Notice that we often had issues with the search API, so instead of just relying on the fact that any user that wasn't found just didn't exist in Auth0, we also stored all users in a DynamoDB Index just to be sure, and checked that as a fallback.
````
search API => Search API

kopertop commented:
Thanks for all the feedback! I really do appreciate a healthy discussion around these issues. I think I addressed everything but feel free to let me know if I missed something. I stuck with the "Email Lookup API" I think I fixed it everywhere; I like how it makes it sound like a larger split from the Search API, since I focused on "Don't use search" a lot here.

kopertop commented:
Was this being merged in or was this post rejected?
